a10networks/a10-octavia
# [STACK-1803][STACK-1805] Add virtual server flavor support. Add flavor key formatting

## Description
Added virtual server flavor and name expression support. Also, included changes to the GetFlavorObject database task to transform `-` to `_` in the key names. This makes it easier to pass the arguments directly the to axapi client in the proceeding create and update tasks.

## Jira Ticket
[STACK-1803](https://a10networks.atlassian.net/browse/STACK-1803)
[STACK-1805](https://a10networks.atlassian.net/browse/STACK-1805)

## Technical Approach
- Refactored the GetFlavorObject name to GetFlavorData
- Modified flows to provide and consume info stored with flavor_data constant key instead of flavor key
- Added parsing of flavor data keys to change from `-` to `_` (ie `arp-disable` to `arp_disable`)
- Repositioned updates to ensure that priority is as follows: name_expr > regular flavor arg > config arg

## Config Changes
N/A

## Test Cases
GetFlavorData Database Task Tests:
- Ensure `_flavor_search` method can find flavor_id from loadbalancer
- Ensure `_flavor_search` method can find flavor_id from listener
- Ensure `_flavor_search` method can find flavor_id from pool
- Ensure `_flavor_search` method can find flavor_id from member
- Ensure `_flavor_search` method can find flavor_id from healthmonitor

- Test that `_format_keys` method works for top-level keys
- Test that `_format_keys` method works for keys inside of lists
- Test that `_format_keys` method works when top-level keys and lists are combined

- Test that when a flavor_id is not found then a None is returned
- Test that when a flavor cannot be found from a flavor_id that a None is returned
- Test that when a flavor does not have an associated flavor_profile then a None is returned
- Test that when all the data is provided then the flavor_data is properly returned

Virtual Server Task Tests:
- Ensure that when creating a virtual server with a flavor the data is sent to the axapi client
- Ensure that when creating a virtual server with a name expression the data is sent to the axapi client

- Test that the flavor data overrides the config data
- Test that the name expression data overrides the config data

## Manual Testing

Step 1: Create flavor profile
```
 openstack loadbalancer flavorprofile create --name lb_test --provider a10 --flavor-data '{"virtual-server": {"name-expressions": [{ "regex": "test", "json": {"arp_disable": 1} }]}}'
```

Step 2:  Create flavor
```
openstack loadbalancer flavor create --name lb_fl_test --flavorprofile lb_test --description "flavor test" --enable
```

Step 3: Create load balancer with the associated flavor
```
openstack loadbalancer create --vip-subnet-id provider-vlan-11-subnet --name test_vip3 --project project_b --flavor lb_fl_test
```

** Note that a load balancer that wasn't created with a name which doesn't have "test" in it will fail to associate the arp_disable but will create the laodbalancer. This is by design. **
